### PR TITLE
Some misc cleanup

### PR DESCRIPTION
**Pull request by naoyam**
- `PairwiseRootDomainMap`: cleaned up option settings
- `IndexSelect, Select, TorchGather, Scatter`: misc cleanups like naming, etc

### CODE REVIEW

File: csrc/ir_internal_nodes.h [hunk 1]
```diff
@@ -58,57 +58,57 @@ class TORCH_CUDA_CU_API SelectOp : public Expr {
...
 
-  virtual const char* getOpString() const override {
+  const char* getOpString() const override {
     return "SelectOp";
   }
 
   std::string toString(int indent_size = 0) const override;
   std::string toInlineString(int indent_size = 0) const override;
 
-  IterDomain* getSelectAxis() const {
```
Comment (naoyam) [inline csrc/ir_internal_nodes.h:77]: No need to store IterDomain as it can be retrieved with `dim()`

File: csrc/root_domain_map.h [hunk 2]
```diff
@@ -93,9 +93,22 @@ class TORCH_CUDA_CU_API PairwiseRootDomainMap : public RootDomainMap {
   //! \param is_exact If true, broadcast andnon-broadcast IDs are not mapped
   explicit PairwiseRootDomainMap(
       const TensorView* producer,
-      const TensorView* consumer,
-      bool is_exact = false,
-      bool require_same_extent = true);
```
Comment (naoyam) [inline csrc/root_domain_map.h:98]: To make option config less confusing, they are now configured with `map***` methods.

File: csrc/root_domain_map.h [hunk 3]
```diff
@@ -117,9 +130,16 @@ class TORCH_CUDA_CU_API PairwiseRootDomainMap : public RootDomainMap {
...
+
+  //! Options to allow more permissive mappings
+
+  //! Map broadcast and non-broadcast domains. Note that this is on by
+  //! default
+  bool map_broadcast_ = true;
+  //! Map domains that may have different extents, e.g., torch_gather
+  bool map_different_extents_ = false;
+  //! Map domains that are indirectly accessed, e.g., index_select
+  bool map_indexed_domains_ = false;
```
Comment (naoyam) [inline csrc/root_domain_map.h:142]: This is a new option. Not used yet.

File: csrc/ir_internal_nodes.h [hunk 4]
```diff
@@ -215,7 +211,7 @@ class TORCH_CUDA_CU_API IotaOp : public Expr {
 
   NVFUSER_DECLARE_CLONE_AND_CREATE
 
-  virtual const char* getOpString() const override {
+  const char* getOpString() const override {
```
Comment (naoyam) [inline csrc/ir_internal_nodes.h:214]: Removed `virtual` as suggested by clang-tidy

**Comment by naoyam** [comment]
!build

File: csrc/ir_nodes.cpp [hunk 5]
```diff
@@ -187,11 +190,7 @@ std::string IndexSelectOp::toString(int indent_size) const {
   indent(ss, indent_size) << output(0)->toString() << "\n";
   indent_size++;
   indent(ss, indent_size) << " = index_select( ";
-  if (input(0)->isA<kir::TensorIndex>()) {
```
Comment (zasdfgbnm) [inline csrc/ir_nodes.cpp:190]: Could you run tests with `$PYTORCH_NVFUSER_DUMP=lower_verbose` to check if this change will be a problem?

File: csrc/root_domain_map.cpp [hunk 6]
```diff
@@ -104,61 +120,70 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
...
+    // no mapping for it.
     if (!squeeze_flags.empty() && squeeze_flags.at(itp)) {
       TORCH_INTERNAL_ASSERT(producer_id->isBroadcast());
       itp++;
       continue;
     }
 
-    // In exact mapping, do not map broadcast domains with
-    // non-broadcast domains
-    if (is_exact_ && producer_id->isBroadcast() != consumer_id->isBroadcast()) {
```
Comment (zasdfgbnm) [inline csrc/root_domain_map.cpp:161]: Thank you for the renaming. I think it makes a lot of sense.

File: csrc/root_domain_map.cpp [hunk 7]
```diff
@@ -104,61 +120,70 @@ std::unordered_map<IterDomain*, IterDomain*> PairwiseRootDomainMap::map(
...
+      // Skip both producer and consumer if mapping not allowed
+      if (!map_indexed_domains_) {
+        itp++;
         itc++;
+        continue;
       }
+    }
+
+    // Condition 2: Different extents
+    if (consumer_tv_->definition()->isA<TorchGatherOp>() &&
```
Comment (zasdfgbnm) [inline csrc/root_domain_map.cpp:158]: nit: in C++17, you can do
```
if (auto gop = dynamic_cast<TorchGatherOp*>(consumer_tv_->definition();  gop != nullptr && gop->exactSizes() && ...)
```
which is slightly less verbose.

**Review by zasdfgbnm — APPROVED** [review]
LGTM, thank you for the cleanup

File: csrc/root_domain_map.cpp (hunk 7 above)
Comment (naoyam) [reply csrc/root_domain_map.cpp:158]: Ha, that's nice. 

I found this article helpful: https://skebanga.github.io/if-with-initializer/

File: csrc/ir_nodes.cpp (hunk 5 above)
Comment (naoyam) [reply csrc/ir_nodes.cpp:190]: That's a good point. I blindly replaced it with `getTvInput` and didn't think why. I don't think it's necessary to print the TensorView of the TensorIndex input. That's not done in any other other `toString` methods. Just changed to call `input(0)->toString()`

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
@zasdfgbnm Looks like there's something wrong (?) with the container, and that's why the CI build is failing. @jjsjann123 is looking into it.

**Comment by zasdfgbnm** [comment]
Hmm, would just retrying again and again help while waiting for the fix? In my PR https://github.com/NVIDIA/Fuser/pull/247, I did a retry, and it helped.

**Comment by naoyam** [comment]
Not sure. I already tried twice.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
Does look like a flaky network issue.
cc'ing @naoyam so I retried this by going to the pipeline link `Test 8148625 completed. Click Details to check pipeline.` and just restarted the failing tests a few times.... this is... unfortunate...

**Comment by zasdfgbnm** [comment]
> Does look like a flaky network issue. cc'ing @naoyam so I retried this by going to the pipeline link `Test 8148625 completed. Click Details to check pipeline.` and just restarted the failing tests a few times.... this is... unfortunate...

Am I reading it correctly that it is a network issue and this issue only happens on this PR? This is super strange. But instead of spending time debugging for this error, should we just use admin hammer to merge?

**Comment by naoyam** [comment]
It doesn't seem just this one since other PRs are also failing #267, #261

**Comment by zasdfgbnm** [comment]
!build

**Comment by jjsjann123** [comment]
There's some status update issue... @zasdfgbnm rerunning `!build` on the same commit actually doesn't really update the status posted on github any more.
https://gitlab-master.nvidia.com/dl/pytorch/update-scripts/-/pipelines/8148625
This was my pipeline before you triggered another build. I do see a few real segfault in the wheel testing... this makes me a little nervous...

**Comment by jjsjann123** [comment]
Yet you current pipeline seems to suggesting the wheel test is fine...

Geee.. these CI nodes are nightmare during code freeze.

**Pull request merged**